src/LibGPS.lua

```diff
@@ -93,15 +93,71 @@ local function GetPlayerWaypoint()
     return LMP:GetMapPing(MAP_PIN_TYPE_PLAYER_WAYPOINT)
 end
 
-local function SetMeasurementWaypoint(x, y)
+local function ClearCurrentWaypoint()
+    currentWaypointX, currentWaypointY = 0, 0, nil
+end
+
+local function RestoreCurrentWaypoint()
+   if(not currentWaypointMapId) then
+        LogMessage(LOG_DEBUG, "Called RestoreCurrentWaypoint without stored waypoint.")
+        return
+    end
+
+    local wasSet = false
+    if (currentWaypointX ~= 0 or currentWaypointY ~= 0) then
+        wasSet = SetPlayerWaypointByWorldLocation(currentWaypointX, 1, currentWaypointY)
+        if (not wasSet) then
+            LogMessage(LOG_DEBUG, "Cannot reset waypoint because it was outside of our reach")
+        end
+    end
+
+    if(wasSet) then
+        LogMessage(LOG_DEBUG, "Waypoint was restored, request pin update")
+        needWaypointRestore = true -- we need to update the pin on the worldmap afterwards
+    else
+        RemovePlayerWaypoint()
+    end
+
+    ClearCurrentWaypoint()
+end
+
+local function SetMeasurementWaypoint(localX, localY)
+    local hasWaypoint = LMP:HasMapPing(MAP_PIN_TYPE_PLAYER_WAYPOINT)
+    if hasWaypoint then
+        currentWaypointX, currentWaypointY = GetPlayerWaypoint()
+        currentWaypointMapId = GetMapTileTexture()
+    end
+
     -- this waypoint stays invisible for others
     lib.suppressCount = lib.suppressCount + 1
     LMP:SuppressPing(MAP_PIN_TYPE_PLAYER_WAYPOINT)
-    LMP:SetMapPing(MAP_PIN_TYPE_PLAYER_WAYPOINT, MAP_TYPE_LOCATION_CENTERED, x, y)
-end
 
-local function SetPlayerWaypoint(x, y)
-    LMP:SetMapPing(MAP_PIN_TYPE_PLAYER_WAYPOINT, MAP_TYPE_LOCATION_CENTERED, x, y)
+    local _, pwx, pwh, pwy = GetUnitWorldPosition("player")
+    local x, y = 2000000, 2000000
+    if not SetPlayerWaypointByWorldLocation(x, pwh, y) then

```

Comment:
Using `SetPlayerWaypointByWorldLocation(1, 1, 1)` should work for all maps. No need for multiple calls with inverted coordinates.

Reply:
Done.